plugins/badges-backend/src/service/moduleContext.ts

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { CatalogClient, catalogApiDep } from '@backstage/catalog-client';
+import { pluginEndpointDiscoveryDep } from '@backstage/backend-common';
+
+import {
+  createDependencyConfig,
+  createDependencyRef,
+} from '@backstage/app-context-common';
+import { BadgeBuilder, DefaultBadgeBuilder } from '../lib';
+import { BadgeFactories } from '../types';
+
+export const badgeFactoriesDep = createDependencyRef<BadgeFactories>(
+  Symbol.for('@backstage/plugin-badges-backend.BadgeFactories'),
+);
+export const badgeBuilderDep = createDependencyRef<BadgeBuilder>(
+  Symbol.for('@backstage/plugin-badges-backend.BadgeBuilder'),
+);
```

Comment:
I realize that there's a benefit from an IDE standpoint to be able to start typing "catalogApiRe..." and getting a completion for where to import it from. But will that be less of a thing in the DI future?

Reply:
Also, this of course has significant spillover onto the ideas of a future plugin architecture for the backend - these exports and needs could make a lot of sense to have as a field of the plugin itself

Reply:
Or how routes are exported, right? https://github.com/backstage/backstage/blob/4c48766266f69ad20d0e819e69064f9e7db546fb/plugins/scaffolder/src/plugin.ts#L59

I think that could make sense, although it would likely need to be in addition to having refs exported by `-node` packages in cases where the intention is to share them for use in other plugins. It's something we could consider for FE as well, to bring down the number of exports.

Reply:
Hm yeah, that's a drawback then. I wouldn't want to have some in -node and some in the main plugin, in that case it's better to always have it in -node. The main thing I was after was to enable the pattern of importing a single symbol whose keys are the refs, always. Shouldn't have to differentiate between `module.exports.foo` and `barRef` from different places depending on what one does.

Reply:
Love it. I think that makes a lot of sense to start creating a better structure around this. Based on your initial code there we can tweak it towards something like:

```
export const module = createDependencyModule({
  defines: {
    badgeFactories: createDependencyRef<BadgeFactories>('@backstage/plugin-badges-backend.BadgeFactories'),
  },
  provides: [
    createDependencyConfig({
      id: catalogModule.defines.catalogApiDep,
      dependencies: { discoveryApi: coreModule.defines.pluginEndpointDiscovery },
      factory: ({ discoveryApi }) => new CatalogClient({ discoveryApi }),
    }),
  ],
  initialize: createRouter,
});
```
Few naming tweaks and addition to have the possibility to expose the main entrypoint to the module in the same structure as well. (And modifying the catalog dep to point to the catalog module, where the interface actually resides).

Regarding IDE support, we'd still get that for the deeper object structure right? In the end the dependency to the actual interface that is the 'exported' one will need to be present, so it makes sense to have the module structure encapsulated to be able to start introducing conventions for these exports/imports.

Reply:
@Xantier We have to be mindful of circular imports though. As @Rugvip alluded to, the `-node` package is the only one that _other_ plugins are able to depend on for their import needs. So it may be hard to put `createRouter`, from the main plugin package, in this declaration because the dependencies are meant to go the other way around.

It may rather be that we have

```ts
// in plugins/foo-backend
import { module as dependencies } from '';
export const plugin = createPlugin({
  dependencies,
  initialize: ...,
  // and default factories for the things declared in the module?
  ...

// in plugins/foo-node
export const module = createDependencyModule({ ...
```

Reply:
I've modified this code and added constructs to export a single `xModule` from each of the defined packages. 
The options here are not really ideal, `definitions` should reside in `-node` packages since `dependencies` will have implementation specific imports within them. 

That meaning that either we export definitions from `-node` as a separate bundle and `dependencies` + entrypoint function from the standard package as a bundle. 

Would that make sense?